### ISSUE TITLE
Limit column or row size upon INSERT or UPDATE

### ISSUE THREAD

**Issue by glennfawcett** [cockroachdb/cockroach#67400]
**Is your feature request related to a problem? Please describe.**
Application was inadvertently INSERTING or UPDATING JSONB columns that were 100MB or larger.  When running batch operations, CockroachDB was aborting with memory errors.  

**Describe the solution you'd like**
It would be nice if CRDB took a defensible position and allowed users to set a limit on the size of a ROW.  This would have avoided problems with erroneous application issues that bloat the size of rows within the database. 

Additional limits that would be useful:

- [x] Limit the number of rows read per query
- [x] Limit the number of inserts/updates in a query
- [x] Limit the row size per table

**Describe alternatives you've considered**
As a workaround for the batch situation, we had to lower the batch size only commit ONE row.  This allowed us to proceed but caused processing delay to diagnosis and resolve.

gz#9005

**Comment by rytaft** [comment]
This will not be possible without increasing raft command size limit (default 64MB), which limits the size of keys in a single transaction. The setting is `kv.raft.command.max_size`. Maybe we need to add a cluster setting that enforces a limit on a single batch? @vy-ton will follow up.

**Comment by michae2** [comment]
According to @sumeerbhola they are setting `kv.raft.command.max_size` to 1GB because of non-deterministic batching of commands.

**Comment by rytaft** [comment]
@glennfawcett @vy-ton -- @michae2 and I were discussing how we might implement limiting the row size, and we had some questions:
- This limit should be on the entire row level, not the individual column family level, correct?
- Should this limit be enforced on all INSERTs, UPDATEs, and UPSERTs?
- What happens if a row that already exists in a table exceeds the limit (e.g., if the limit was added after the row was inserted)? Should any attempt to UPDATE that row fail, unless it reduces the size of the row below the limit?

**Comment by glennfawcett** [comment]
Yes to all above...

**Comment by michae2** [comment]
@glennfawcett @vy-ton -- @rytaft and I were talking more and wondering, would something like this fit the bill?

```
ALTER TABLE x ADD CONSTRAINT x_row_size_limit CHECK (pg_column_size(*) < limit)
```

This would be a per-table limit that would be enforced on all modifications. There would be a validation step when first applied to ensure all existing rows are within the limit.

**Comment by vy-ton** [comment]
A potential issue here is that the `ADD CONSTRAINT` option would require the AppDev teams to make changes in order to opt-in to the guardrail. This is unlikely something the DBaaS could enforce for the whole cluster.

**Comment by RaduBerinde** [comment]
Note that implementing something like this for UPDATE/UPSERT can significantly affect workloads if column families are in use. We would have to always read the entire row to check the limit, even if otherwise we'd be touching a single column family. This would remove the contention-related benefit of column families.

**Comment by michae2** [comment]
Based on a call with @vy-ton it sounds like at first the row-size limit should be a cluster setting, rather than a per-table setting. And based on @RaduBerinde's excellent point above, this probably needs to be a column-family-size limit rather than a full-row-size limit.

**Comment by vy-ton** [comment]
If a table doesn't have column families, would a column family size limit then be equal to the whole row?

Other options we can validate with users:

- The full row limit could be ignored for tables with column families

**Comment by michae2** [comment]
Yes, if a table doesn't have explicit column families then it has one implicit column family which consists of all of the columns.

**Comment by michae2** [comment]
It looks like https://github.com/cockroachdb/cockroach/pull/66927 might address limiting the number of rows modified in a query.

**Comment by michae2** [comment]
This is now fixed in 21.1, 21.2, and master.
